content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+import MockedMap from "./map";
+
+jest.mock("./map", () => jest.fn(() => "DummyMap"));
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("should render contact information", () => {
+  const center = { lat: 0, lang: 0 };
+  act(() => {
+    render(<Contact name="" email="" site="" center={center} />, container);
+  });
+  // ensure the mocked map component function was called correctly
+  expect(MockedMap).toHaveBeenCalledWith({ id: "example-map", center }, {});
+});
+```
+
+### Events {#events}
+
+Dispatch real events on components and elements, and observe the side effects of those actions on UI elements. Consider a `Toggle` component:
+
+```jsx
+// toggle.js
+import React, { useState } from "react";
+function Toggle(props) {
+  // initial, onChange
+  const [state, setState] = useState(Boolean(initial));
+  return (
+    <button
+      onClick={() => {
+        setState(previousState => !previousState);
+        onChange(!previousState);
+      }}
+      data-testid="toggle"
+    >
+      {state === true ? "Turn off" : "Turn on"}
+    </button>
+  );
+}
+```
+
+We could write tests for it:
+
+```jsx
+// toggle.test.js
+import React from "react";
+import { render } from "react-dom";
+import Toggle from "./toggle";
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+it("changes value when clicked", () => {
+  // a counter to track clicks
+  let clicks = 0;
+  act(() => {
+    render(<Toggle initial={true} onChange={() => clicks++} />, container);
+  });
+  // get a hold of the button element, and trigger some clicks on it
+  const button = document.querySelector("[data-testid=toggle]");
+  expect(button.innerText).toBe("Turn on!");
+  act(() => {
+    button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+  });
+  expect(clicks).toBe(1);
+  expect(button.innerText).toBe("Turn off!");
+  act(() => {
+    for (let i = 0; i < 5; i++) {
+      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+    }
+  });
+  expect(clicks).toBe(6);
+  expect(button.innerText).toBe("Turn on!");
+});
+```
+
+### Timers {#timers}
+
+UIs could use timer based functions like `setTimeout` to run time sensitive code. In this example, a multiple choice panel waits for a selection and advances, timing out if a selection isn't made in 5 seconds.
+
+```jsx
+function Card(props) {
+  useEffect(
+    props => {
+      const timeoutID = setTimeout(() => {
+        props.onSelect(null);
+      }, 5000);
+      return () => {
+        clearTimeout(timeoutID);
+      };
+    },
+    [props.onSelect]
```

Comment:
This doc uses Hooks for everything. None of other docs outside Hooks section do. Will this give an impression that these examples don't work for classes? Maybe we should have a mix here.

Reply:
Instead of a mix, a single sentence that says: "Note: These testing strategies work just as well for class components as they do for function components."